kvserver: auto-tune closed timestamp

wenyihu6 opened the pull request:
Note that this PR is based on top of https://github.com/cockroachdb/cockroach/pull/142664.

---

**kvserver/closedts: plumb latency tracker**

This commit integrates the latency tracker with the server, side transport
sender, and store. While it currently doesn't do anything, future commits will
use it to estimate the closed timestamp target.

Epic: none
Release note: none

---

**kvserver/closedts: add latency tracker to side transport**

For global table ranges, we aim to ensure follower replicas can serve
present-time reads by calculating a target closed timestamp that leads the
current time (using the `LEAD_FOR_GLOBAL_READS` policy). The goal of this
estimation is to be confident that by that timestamp that the all follower
replicas should have received this closed timestamps and can serve present time
reads without redirecting to the leaseholders. Note that this is an estimated
goal, not the exact lead time. It is still possible that some follower nodes are
really behind or we fail to bump closed timestamps.

Calculation:
raft_propagation_time = max_network_rtt*1.5 + raft_overhead(20ms)
side_propagation_time = max_network_rtt*0.5 + side_transport_close_interval(200ms by default)
closed_ts_propagation = max(raft_propagation_time,side_propagation_time)
closed_ts_at_sender = now(sender’s current clock) + maxClockOffset(500ms by
default) + buffer (25ms) + closed_ts_propagation

Currently, `max_network_rtt` is hardcoded at 150ms, leading to a default 800ms lead.

Problems with the current implementation:

Tuning trade-offs: A higher lead timestamp forces leaseholders to wait longer
for the current time to catch up before returning to client, incurring a high
write latency. A lower lead timestamp may result in followers not fully
replicating the changes before returning control to the client, forcing follower
replicas to redirect query to the leaseholder and increase read latency.

Cluster wide setting: kv.closed_timestamp.lead_for_global_reads_override only
applies to the entire cluster, not per range. For geographically distant
replicas, a short lead time may not be enough for full application. However,
increasing the lead time cluster wide can harm write performance for replicas
that are close together and don’t require as much time to replicate.

This commit adds a latency tracker to the side transport senders.

Currently, nodes create a map of closed timestamps for each policy without
consulting leaseholders. Senders pass this map to leaseholders, who decide
whether to "bump" their closed timestamp and which policy to use. Senders then
send this information to receivers to opt in ranges and pick the closed
timestamps based on the policies ranges opt in.

Currently there are only two policies in the map side transport senders create.
We lack fine grained control into the closed timestamps based on the ranges. The
new design adds three more policies based on range locality information:
cross-region, cross-zone (same region), and single-region.
Every range would be broken down into 4 categories:
1. No locality - fall back to previous if no auto-tuning is enabled or no
locality information is found for ranges
2. Cross region
3. Cross zone, same region
4. Same region, same zone

Periodically, side transport determines the latency for these 4 categories based
on its open connections. Note that this may not be 100% accurate since the set
of follower nodes may be changing, but it will be close enough. Side transport
then consults with its leaseholders. Leaseholders decide each range’s category
based on its replicas. This result is cached for raft side computation.
Leaseholders also use the category cached above and check with the latency
tracker to get the network latency as well when computing a closed timestamp
target for raft log propagation.

kv.closed_timestamp.lead_for_global_reads_auto_tune.enabled can now be used
to auto-tune the lead time that ranges global tables use to publish
close timestamps. The kv.closed_timestamp.lead_for_global_reads_override
cluster setting takes precedence over this one. If auto-tuning is disabled or
no data is observed, the system falls back to a hardcoded computed lead time.

Resolves: #59680
Release note: none

wenyihu6 committed:
kvserver: add kv.closed_timestamp.lead_for_global_reads_auto_tune_interval

This commit adds a new cluster setting,
`kv.closed_timestamp.lead_for_global_reads_auto_tune_interval` to control how
frequently the system measures latency between nodes. If non-zero, the observed
latency between nodes in the cluster would be used to dynamically auto-tune
the lead time for ranges using the LEAD_FOR_GLOBAL_READS policy.
The setting defaults to 5 minutes and can be disabled by setting to 0.
It only takes effect when `side_transport_interval` is non-zero.

Note that setting is currently unused. Future commits will populate and
integrate it.

Part of: #59680
Release note: none

wenyihu6 committed:
kvserver/closedts: change TargetForPolicy to take network rtt

This patch changes TargetForPolicy to take in network round trip latency.
Currently, it always takes in closedts.DefaultMaxNetworkRTT. But future commits
will change it to take in a dynamically observed network latency.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142399)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver/closedts: add acceptable rtt range to TargetForPolicy

This patch changes TargetForPolicy to clamp the latency taken in within an
acceptable latency range. Currently, it always takes in
closedts.DefaultMaxNetworkRTT. But future commits will change it to take in a
dynamically observed network latency.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: replace roachpb.RangeClosedTimestampPolicy with RangeClosedTimestampByPolicyLocality

This commit replaces roachpb.RangeClosedTimestampPolicy with
ctpb.RangeClosedTimestampByPolicyLocality. Currently, it maintains the existing
behavior. Future commits will introduce additional enum values to add locality
information into the side-transport closed timestamp estimation.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: add latency tracker to side transport

For global table ranges, we aim to ensure follower replicas can serve
present-time reads by calculating a target closed timestamp that leads the
current time (using the `LEAD_FOR_GLOBAL_READS` policy). The goal of this
estimation is to be confident that by that timestamp that the all follower
replicas should have received this closed timestamps and can serve present time
reads without redirecting to the leaseholders. Note that this is an estimated
goal, not the exact lead time. It is still possible that some follower nodes are
really behind or we fail to bump closed timestamps.

Calculation:
raft_propagation_time = max_network_rtt*1.5 + raft_overhead(20ms)
side_propagation_time = max_network_rtt*0.5 + side_transport_close_interval(200ms by default)
closed_ts_propagation = max(raft_propagation_time,side_propagation_time)
closed_ts_at_sender = now(sender’s current clock) + maxClockOffset(500ms by
default) + buffer (25ms) + closed_ts_propagation

Currently, `max_network_rtt` is hardcoded at 150ms, leading to a default 800ms lead.

Problems with the current implementation:

Tuning trade-offs: A higher lead timestamp forces leaseholders to wait longer
for the current time to catch up before returning to client, incurring a high
write latency. A lower lead timestamp may result in followers not fully
replicating the changes before returning control to the client, forcing follower
replicas to redirect query to the leaseholder and increase read latency.

Cluster wide setting: kv.closed_timestamp.lead_for_global_reads_override only
applies to the entire cluster, not per range. For geographically distant
replicas, a short lead time may not be enough for full application. However,
increasing the lead time cluster wide can harm write performance for replicas
that are close together and don’t require as much time to replicate.

This commit adds a latency tracker to the side transport senders.

Currently, nodes create a map of closed timestamps for each policy without
consulting leaseholders. Senders pass this map to leaseholders, who decide
whether to "bump" their closed timestamp and which policy to use. Senders then
send this information to receivers to opt in ranges and pick the closed
timestamps based on the policies ranges opt in.

Currently there are only two policies in the map side transport senders create.
We lack fine grained control into the closed timestamps based on the ranges. The
new design adds three more policies based on range locality information:
cross-region, cross-zone (same region), and single-region.
Every range would be broken down into 4 categories:
1. No locality - fall back to previous if no auto-tuning is enabled or no
locality information is found for ranges
2. Cross region
3. Cross zone, same region
4. Same region, same zone

Periodically, side transport determines the latency for these 4 categories based
on its open connections. Note that this may not be 100% accurate since the set
of follower nodes may be changing, but it will be close enough. Side transport
then consults with its leaseholders. Leaseholders decide each range’s category
based on its replicas. This result is cached for raft side computation.
Leaseholders also use the category cached above and check with the latency
tracker to get the network latency as well when computing a closed timestamp
target for raft log propagation.

kv.closed_timestamp.lead_for_global_reads_auto_tune.enabled can now be used
to auto-tune the lead time that ranges global tables use to publish
close timestamps. The kv.closed_timestamp.lead_for_global_reads_override
cluster setting takes precedence over this one. If auto-tuning is disabled or
no data is observed, the system falls back to a hardcoded computed lead time.

Resolves: #59680
Release note: none

wenyihu6 committed:
kvserver/closedts: add latency tracker

This commit adds a latency tracker to measure network latency between nodes
based on locality. Note that it is currently unused. Future commits will update
the side transport sender to periodically refresh latency metrics and
incorporate them into closed timestamp estimation.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: pull test helpers into sender_helper_test.go

This commit extracts test helpers into a separate file, sender_helper_test.go.
While it doesn't alter existing behavior, this change prepares for future
commits that will introduce additional helper utilities and tests which make
this file too large.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: add policy locality

This commit introduces GetTargetAndPolicyLocality, a utility for the
sidetransport.Replica interface that helps determine which targetByPolicy
replicas to select during BumpSideTransportClosed, based on the replicas'
localities and policies.

The utility is placed under sidetransport to simplify testing, as sidetransport
sender tests use mock replicas which will exercise this function to help facilitate
testing in future commits.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: plumb latency tracker

This commit integrates the latency tracker with the server, side transport
sender, and store. While it currently doesn't do anything, future commits will
use it to estimate the closed timestamp target.

Epic: none
Release note: none
